tools/testsys/src/wait.rs

```diff
@@ -0,0 +1,285 @@
+                .allow_not_found(|_| warn!("A resource could not be found."))
+                .context(error::TestsysClientSnafu)
+                .transpose()
+        })
+        .collect::<Result<Vec<Resource>>>();
+
+        match resources {
+            resources @ Ok(_) => break resources,
+            Err(_) if count < n_iters => {
+                count += 1;
+                warn!("Unable to fetch resources, retrying {count} of {n_iters}");
+            }
+            error @ Err(_) => break error,
+        }
+    }
+}
+
+/// Try to fetch the names of all tests that are dependent on this resource.
+async fn fetch_tests_using_resource(
+    client: &TestManager,
+    resource_name: &str,
+) -> Result<Vec<Test>> {
+    let test_client = client.test_client();
+    let mut count = 0;
+    let n_iters = 5;
+    let tests = loop {
+        let tests = test_client.get_all().await;
+        match tests {
+            tests @ Ok(_) => break tests,
+            Err(_) if count < n_iters => {
+                count += 1;
+                warn!("Unable to fetch all tests, retrying {count} of {n_iters}");
+            }
+            error @ Err(_) => break error,
+        }
+    }
+    .context(error::TestsysClientSnafu)?;
+
+    Ok(tests
+        .iter()
+        .filter(|test| test.spec.resources.contains(&resource_name.to_string()))
+        .cloned()
+        .collect())
+}
+
+/// Wait until the test has completed.
+async fn wait_for_test(client: &TestManager, test_name: &str) -> Result<TestRunResults> {
+    let test_client = client.test_client();
+    loop {
+        let test = test_client
+            .get(test_name)
+            .await
+            .context(error::TestsysClientSnafu)?;
+
+        match test.agent_status().task_state {
+            TaskState::Running | TaskState::Unknown => {
+                info!("Test '{test_name} is still running. Sleeping 60s.")
+            }
+            TaskState::Completed => {
+                return Ok(TestRunResults::Successful {
+                    results: vec![(test_name.to_string(), test.agent_status().results.clone())]
+                        .into_iter()
+                        .collect(),
+                })
+            }
+            TaskState::Error => {
+                return Ok(TestRunResults::TestError {
+                    test: test_name.to_string(),
+                    error: test.agent_error().unwrap_or_default().to_string(),
+                })
+            }
+        }
+        tokio::time::sleep(Duration::from_secs(60)).await;
+    }
+}
+
+/// Wait until the CRD has completed and determine whether it was successful or not.
+async fn wait_for_crd(client: &TestManager, crd: &Crd) -> Result<TestRunResults> {
+    match crd {
+        Crd::Test(test) => wait_for_test(client, test.object_name()).await,
+        Crd::Resource(resource) => wait_for_resource_creation(client, resource.object_name()).await,
+    }
+}
+
+#[derive(Debug)]
+pub enum TestRunResults {
```

Comment:
A plural noun implies some sort of collection.

```suggestion
pub enum TestRunResult {
```

Reply:
> There should be a configurable overall timeout, i.e. we might want the --wait invocation to timeout after 3 hours instead of hanging indefinitely.

We definitely don't want a default overall timeout. A sonobuoy test that is waiting for another sonobuoy test to finish would certainly take more than 3 hours. I'll add a timeout for waiting for other completed resources to be deleted (10 mins). I will also add an optional overall timeout.